test/runnable/test19386.d

```diff
@@ -0,0 +1,36 @@
+struct Thing
+{
+    this(int* i)
+    {
+        ptr = i;
+        (*ptr)++;
+    }
+
+    ~this()
+    {
+        (*ptr)--;
+    }
+
+    T opCast(T : bool)()
+    {
+        return false;
+    }
+
+    int* ptr;
+}
+
+Thing makeThing(int* p)
+{
+    return Thing(p);
+}
+
+void main()
+{
+    int i;
+    {
+        if (auto t = makeThing(&i)) // destructor not called
+        {
+        }
```

Comment:
An additional test that also asserts that `i` is one in an `else` block would be nice too.

Reply:
you mean that `i == 0`, as per the [spec](https://dlang.org/spec/statement.html#IfStatement) _"Its scope extends from when it is initialized to the end of the ThenStatement"_.

Reply:
Yes, you are right. I misled myself by mixing the generated code in the PR description with actual language semantics.

Reply:
To accomplish what I want to be tested we need to have two separate variables counting the number of times the constructor and destructor were called.

Reply:
Why the constructor though? Why check if the constructor is called more than once?